File: CPP/2010-09/task4.cc
#include <iostream>

using namespace std;

class Stack {

private:

    const static int CHUNK = 2;

    int size;
    char* data;
    int top;

public:

    Stack():
    size(CHUNK),
    data(new char[CHUNK]),
    top(-1)
    {}

    ~Stack() {
        delete [] data;
    }

    void push(char value) {
        if (top >= (size-1)) {
            resize();
        }

        data[++top] = value;
    }

    char pop() {
        if (top < 0) {
            cout << "Stack is empty" << endl;
            return ' ';
        }

        return data[top--];
    }

    char head() {
        if (top < 0) {
            cout << "Stack is empty" << endl;
            return ' ';
        }

        return data[top];
    }

    void resize() {
        cout << "Resize called" << endl;

        char *temp = data;
        data = new char[size + CHUNK];

        for (int i=0; i<size; i++) {
            data[i] = temp[i];
        }

        delete [] temp;
        size += CHUNK;
    }

    bool is_empty() {
        return top == -1;
    }

};

bool validate_brackets_string(string str) {
    Stack stack;

    for (int i=0; i < str.length(); ++i) {
        char current = str[i];

        if (current == '(') {
            stack.push(current);
        }
        else if (current == ')') {
            if (stack.head() == '(') {
                stack.pop();
            }
            else {
                return false;
            }
        }
    }

    return stack.is_empty();
}

int main(int argc, char** argv) {

    // Stack stack = Stack();

    // cout << "PUSH: (" << endl;
    // stack.push('(');
    // cout << "PUSH: )" << endl;
    // stack.push(')');
    // cout << "PUSH: )" << endl;
    // stack.push(')');

    // cout << "POP: " << stack.pop() << endl;
    // cout << "POP: " << stack.pop() << endl;
    // cout << "POP: " << stack.pop() << endl;

    string validInput1 = "((()))";
    string validInput2 = "()()()";
    string validInput3 = "(()())((()()))";

    string invalidInput1 = "((()";
    string invalidInput2 = "(((";
    string invalidInput3 = ")))";
    string invalidInput4 = ")()()(";
    string invalidInput5 = "()())";

    cout << validInput1 << ": " << validate_brackets_string(validInput1) << endl;
    cout << validInput2 << ": " << validate_brackets_string(validInput2) << endl;
    cout << validInput3 << ": " << validate_brackets_string(validInput3) << endl;

    cout << invalidInput1 << ": " << validate_brackets_string(invalidInput1) << endl;
    cout << invalidInput2 << ": " << validate_brackets_string(invalidInput2) << endl;
    cout << invalidInput3 << ": " << validate_brackets_string(invalidInput3) << endl;
    cout << invalidInput4 << ": " << validate_brackets_string(invalidInput4) << endl;
    cout << invalidInput5 << ": " << validate_brackets_string(invalidInput5) << endl;

}

File: Java/src/е2010_09/Task4.java
package е2010_09;

class Node {
	Node next;
	Node prev;
	char data;
	
	Node(char value) {
		next = null;
		prev = null;
		data = value;
	}
}

class Stack {
	
	Node first;
	Node last;
	
	Stack() {
		first = null;
		last = null;
	}
	
	void push(char value) {
		Node node = new Node(value);
		appendElement(node);
	}
	
	char pop() {
		Node node = removeLastElement();
		if (node == null) {
			return ' ';
		}
		
		return node.data;
	}
	
	void appendElement(Node node) {
		if (last == null) {
			first = node;
			last = node;
		}
		else {
			last.next = node;
			node.prev = last;
			last = node;
		}
	}
	
	Node removeLastElement() {
		if (isEmpty()) {
			return null;
		}
		
		if (last == first) {
			Node node = last;
			
			last = null;
			first = null;
			
			return node;
		}
		
		Node node = last;
		last = node.prev;
		last.next = null;
		
		node.prev = null;
		node.next = null;
		
		return node;
	}
	
	boolean isEmpty() {
		return last == null;
	}
	
}

public class Task4 {
	
	public static boolean validateBracketsString(String input) {
		Stack stack = new Stack();
		
		for (int i=0; i < input.length(); ++i) {
			if (input.charAt(i) == '(') {
				stack.push(input.charAt(i));
			}
			else if (input.charAt(i) == ')') {
				if (stack.pop() != '(') {
					return false;
				}
			}
		}
		
		return stack.isEmpty();
	}

	public static void main(String[] args) {
		
//		Stack stack = new Stack();
//		
//		stack.push('(');
//		stack.push(')');
//		
//		System.out.println("POP: " + stack.pop());
//		System.out.println("POP: " + stack.pop());
//		System.out.println("POP: " + stack.pop());

		String validInput1 = "((()))";
	    String validInput2 = "()()()";
	    String validInput3 = "(()())((()()))";

	    String invalidInput1 = "((()";
	    String invalidInput2 = "(((";
	    String invalidInput3 = ")))";
	    String invalidInput4 = ")()()(";
	    String invalidInput5 = "()())";
		
		System.out.println(validInput1 + " -> " + validateBracketsString(validInput1));
		System.out.println(validInput2 + " -> " + validateBracketsString(validInput2));
		System.out.println(validInput3 + " -> " + validateBracketsString(validInput3));
		
		System.out.println(invalidInput1 + " -> " + validateBracketsString(invalidInput1));
		System.out.println(invalidInput2 + " -> " + validateBracketsString(invalidInput2));
		System.out.println(invalidInput3 + " -> " + validateBracketsString(invalidInput3));
		System.out.println(invalidInput4 + " -> " + validateBracketsString(invalidInput4));
		System.out.println(invalidInput5 + " -> " + validateBracketsString(invalidInput5));
		
	}

}
